src/import_edf.jl

```diff
@@ -208,81 +302,264 @@ function groupby(f, list)
+        return rowmerge(row; edf_signal_idx)
+    end
 
-`unit_alternatives` lists standardized unit names and alternatives that map to them.
-See `OndaEDF.STANDARD_UNITS` for defaults.
+    # group signals by which Samples they will belong to, promote_encoding, and
+    # write index of destination signal into plan to capture grouping
+    grouped_rows = groupby(onda_signal_groups, plan_rows)
+    plan_rows = mapreduce(vcat, enumerate(values(grouped_rows))) do (onda_signal_idx, rows)
+        encoding = promote_encodings(rows)
+        return [rowmerge(row, encoding, (; onda_signal_idx)) for row in rows]
+    end
 
-`preprocess_labels(label::String, transducer_type::String)` is applied to raw edf signal header labels
-beforehand; defaults to returning `label`.
+    return FilePlan.(plan_rows)
+end
 
-See `OndaEDF.STANDARD_LABELS` for the labels (`signal_names => channel_names`
-`Pair`s) that are used to extract EDF signals by default.
+_get(x, property) = hasproperty(x, property) ? getproperty(x, property) : missing
+function grouper(vars=(:kind, :sample_unit, :sample_rate))
+    return x -> NamedTuple{vars}(_get.(Ref(x), vars))
+end
 
+# return Samples for each :onda_signal_idx
 """
-function extract_channels_by_label(edf::EDF.File, signal_names, channel_names;
-                                   unit_alternatives=STANDARD_UNITS, preprocess_labels=(l,t) -> l)
-    matcher = x -> begin
-        # yo I heard you like closures
-        # x is either a channel name (string), or a channel_name => alternatives Pair.
-        this_channel_name = x isa Pair ? first(x) : x
-        return s -> match_edf_label(preprocess_labels(s.header.label, s.header.transducer_type),
-                                    signal_names,
-                                    this_channel_name,
-                                    channel_names)
-    end
-    edf_channels = extract_channels(edf.signals, (matcher(x) for x in channel_names))
+    edf_to_onda_samples(edf::EDF.File, plan_table; validate=true,
+                        samples_groups=grouper((:onda_signal_idx, )))
 
-    # place channels with different physical units or sample rates in separate Onda signals
-    grouped = groupby(edf_channels) do p
-        channel_name, edf_signal = p
-        return ((edf_signal.header.samples_per_record / edf.header.seconds_per_record), edf_signal.header.physical_dimension)
-    end
+Convert Signals found in an EDF File to `Onda.Samples` according to the plan
+specified in `plan_table` (e.g., as generated by [`plan_edf_to_onda_samples`](@ref)), returning an
+iterable of the generated `Onda.Samples` and the plan as actually executed.
 
-    results = map(values(grouped)) do pairs
-        # pairs is a vector of `standard_onda_name::String => EDF.Signal` pairs
-        edf_channel_names, edf_channels = zip(pairs...)
+The input plan is transformed by using [`merge_samples_info`](@ref) to combine
+rows with the same `:onda_signal_idx` (or output of `sample_groups`) into a
```

Comment:
What's an example of a case where you might want to group by something other than the Onda signal index, i.e. the grouping already determined in `plan_table`? IMO this seems like a weird thing to support but perhaps I'm not seeing its  f u l l   p o t e n t i a l.

Reply:
Yeah, I can imagine a situation where someone might want to split two things that are grouped together into separate groups but not re-compute the indices.  While that might be better handled by custom labels, I don't really see a downside to `reverting control to the caller` here

Reply:
I'm confused. If someone wants to group things differently then aren't the previously computed indices already irrelevant?

Reply:
they might be MOSTLY correct but need a little tweak.  you can then insert another index column that gets combined with the original index, so e.g. everywhere 1 but a few 2s to separate out a few channels into distinct samples from the other channels with the same `onda_signal_index`,

Reply:
that might be a good test case actually...

Reply:
I'm actually coming around to the idea that there should be no option to change which column gets used for this, but that we should refactor and possibly export the function that actually does the grouping/index writing, which would make it more ergonomic to work with the single-channel `plan` interface.  theeeennnn maybe we can talk about getting rid of the preprocessor :)

Reply:
I've now refactored teh grouping stuff into its own function `plan_edf_to_onda_signal_groups` which takes a plan table and a (tuple/vector of) columns to groupby.  this will check whether the `edf_signal_index` column already exists (which it might not if someone has run `plan_edf_to_onda_samples` one signal at a time); if it doesn't it gets filled in based on the order of the rows passed in.  This introduces the possibility of errors if the rows have been re-ordered, but there is already a check (with a nice error message) that the signal label matches the plan label when doing the actual conversion into samples so I'm comfortable that that won't cause big problems.